README
NAME
    File::Tempdir::ForPackage - Easy temporary directories associated with
    packages.

VERSION
    version 1.000002

DESCRIPTION
    This is mostly an interface wrapper for "File::Temp::tempdir", stealing
    code from "File::Tempdir";

    *   I constantly forget how "File::Tempdir" works

    *   I often want a "tempdir" with the name of the package working with
        it enshrined in the path

    *   I constantly forget the magic glue syntax to get a folder inside a
        system "tempdir" with a custom prefix and a user defined length of
        random characters.

    And this is designed to solve this simply.

    use File::TempDir::ForPackage;

      my $tempdir = File::TempDir::ForPackage->new( package => __PACKAGE__ , use_version => 1 );
      my $dir = $tempdir->dir();

    do shit in `$dir` $dir on Linux will be something like
    /tmp/perl-Some-Package-maybewith-a-VERSION-AFG14561/ so if it crashes
    and leaves a "tempdir" behind, you will know who left that "tempdir"
    behind and have a way of cleaning it up.

    When $tempdir is destroyed, $dir will be cleaned;

    Additionally:

      $dir->run_once_in(sub{
        ...
      });

    Is there for people who don't trust scope auto-cleansing and want to
    know when the directory is reaped.

    Additionally, this code can be run in a tight loop creating and
    destroying lots of similarly named temporary directories without risk of
    conflict.

      for my $i ( 0 .. 30  ) {
        $dir->run_once_in(sub {
          system 'find $PWD';
        });
      }

    This emits something like:

      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-PzH4BD
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-5h8nkG
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-UXKt4S
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-Lqg2aW
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-DkNeq6
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-jRI_zF
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-j0_Gt1
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-iX1ddT
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-ZmvikK
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-QNGOUF
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-6wssvL
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-ZmwZxl
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-wIzRTs
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-xetCym
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-8Y0vyX
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-Zlqt6X
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-U5Z_Sa
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-sKmow1
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-rUND95
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-XjPSGF
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-ec8sZZ
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-_4NBwX
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-xM9i6l
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-p3FhJf
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-Zv0sso
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-rP8cAi
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-iade0x
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-fsDDPy
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-FeCcfZ
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-ta5yfg
      /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-rdcQhF

    Except of course, with a package of your choosing, and possibly that
    packages version.

METHODS
  "preserve"
    Toggle the preservation of the "tempdir" after it goes out of scope or
    is otherwise indicated for cleaning.

      $instance->preserve(); # tempdir is now preserved after cleanup
      $instance->preserve(0); # tempdir is purged at cleanup
      $instance->preserve(1); # tempdir is preserved after cleanup

    Note that in "run_once_in", a new "tempdir" is created and set for this
    modules consumption for each run of "run_once_in", regardless of this
    setting. All this setting will do, when set, will prevent each instance
    being reaped from the file system.

    Thus:

      $dir->preserve(1);
      for( 1..10 ){
        $dir->run_once_in(sub{

        });
      }

    Will create 10 temporary directories on your file system and not reap
    them.

  "dir"
    Return a path string to the created temporary directory

      my $path = $instance->dir

  "run_once_in"
    Vivifies a temporary directory for the scope of the passed sub.

      $instance->run_once_in(sub{
        # temporary directory is created before this code runs.
        # Cwd::getcwd is now inside the temporary directory.
      });

      # temporary directory is reset, and possibly reaped.

    You can call this method repeatedly, and you'll get a separate temporary
    directory each time.

ATTRIBUTES
  "package"
    The package to report as being associated with. This really can be any
    string, as its sanitized and then used as a path part.

    If not specified, will inspect "caller"

      my $instance = CLASS->new(
        package => 'Something::Here',
        ...
      );

    Note: If you want "with_version" to work properly, specifying a valid
    package name will be helpful.

  "with_version"
    Include the version from "package->VERSION()" in the "tempdir" path.

    Defaults to false.

      my $instance = CLASS->new(
        ...
        with_version => 1,
      );

  "with_timestamp"
    Include "time" in the "tempdir" path.

    Defaults to false.

      my $instance = CLASS->new(
        ...
        with_timestamp => 1,
      );

  "with_pid"
    Include $$ in the "tempdir" path.

    Defaults to false.

      my $instance = CLASS->new(
        ...
        with_pid => 1,
      );

  "num_random"
    The number of characters of randomness to include in the "tempdir"
    template.

    Defaults to 8. Must be no lower than 4.

      my $instance = CLASS->new(
        ...
        num_random => 5,
      );

PRIVATE ATTRIBUTES
  "_preserve"
    Internal "boolean" for tracking the _preserve state.

  "_dir"
    Internal "File::Tempdir" path.

PRIVATE METHODS
  "_build__dir"
    Builder method for _dir which creates a temporary directory based on the
    passed parameters.

PRIVATE FUNCTIONS
  "_clean_pkg"
    Scrape garbage out of the 'package' field for use in file system tokens.

  "_clean_ver"
    Scrape garbage out of versions for use in file system tokens.

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for File-Tempdir-ForPackage

1.000003 2017-03-10T11:36:17Z d6e15be

1.000002 2017-03-10T11:35:18Z 2b67ff5
 - Avoid test failures due to -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

1.000001 2014-08-15T22:56:28Z
 [00 Trivial]
 - CPANDAY!
 - no code changes
 - whitespace reduced.

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - develop: ↑1 (suggests: ↑1)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

1.000000 2014-08-14T08:50:45Z
 [00 Major]
 - Changes that will affect hand operated downstreams.

 [01 Version Scheme]
 - Migrated to x.yyyzzz from x.y.z
 - This is the primary reason for the 1.x on the box.
 - https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129

 [02 Tooling]
 - Moved to ExtUtils::MakeMaker
 - "Latest" deps for toolchain parts are now only "recommended"
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [03 Cleanup functions nuked from orbit]
 - The persist function is now handled on a tempdir basis by directly using File::Temp tricks.
 - This means that any tempdir that is to be freed will be freed as soon as it is no longer referenced
 - so ->_clear_dir will wipe it
 - and program termination should wipe it.
 - so cleanse and DESTROY are no longer needed

 [Dependencies::Stats]
 - Dependencies changed since 0.1.1, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +49 -1 (recommends: -1, suggests: +1 ↑1)
 - runtime: +2 ↓1 -2
 - test: +2 ↓1 -1 (recommends: +3)

0.1.1 2012-07-26T14:43:40Z
 [Documentation]
 - Full documentation of all directly defined properties.

0.1.0 2012-07-26T12:58:43Z
 - First version.


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4206"
  },
  "configure_requires" => {
    "Module::Build" => "0.4206"
  },
  "dist_abstract" => "Easy temporary directories associated with packages.",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "File-Tempdir-ForPackage",
  "dist_version" => "0.1.2",
  "license" => "perl",
  "module_name" => "File::Tempdir::ForPackage",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "File::Path" => 0,
    "File::Temp" => 0,
    "File::pushd" => 0,
    "Moo" => 0,
    "Sub::Quote" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "1.001003"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Build" => "0.4206",
  "Test::More" => "1.001003"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Easy temporary directories associated with packages.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "File-Tempdir-ForPackage",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "File::Tempdir::ForPackage",
  "PREREQ_PM" => {
    "Carp" => 0,
    "File::Temp" => 0,
    "File::pushd" => 0,
    "Moo" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.000003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "File::pushd" => 0,
  "Moo" => 0,
  "Test::More" => 0,
  "constant" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
$bundle->remove_policy('ValuesAndExpressions::ProhibitConstantPragma');

#$bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



lib/File/Tempdir/ForPackage.pm
use 5.006;
use strict;
use warnings;

package File::Tempdir::ForPackage;

our $VERSION = '1.000003';

# ABSTRACT: Easy temporary directories associated with packages.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Carp qw( croak );
use Moo qw( has );
use File::Temp qw();
## no critic (Subroutines::ProhibitCallsToUnexportedSubs)
use constant MINX => File::Temp::MINX;
## use critic

















has package => (
  is   => ro =>,
  lazy => 1,
  default => sub { scalar [ caller 1 ]->[0] },
);


















































has with_version   => ( is => ro =>, lazy => 1, default => sub { undef } );
has with_timestamp => ( is => ro =>, lazy => 1, default => sub { undef } );
has with_pid       => ( is => ro =>, lazy => 1, default => sub { undef } );
has num_random     => (
  is  => 'ro',
  isa => sub {
    return if $_[0] >= MINX();
    croak( "num_random ( $_[0] ) must be >= " . MINX() );
  },
  default => sub { 8 },
);







has '_preserve' => ( is => rw =>, init_arg => 'preserve', lazy => 1, default => sub { 0 } );







has _dir => ( is => 'lazy', clearer => 1, predicate => 1 );

























sub preserve {
  my ( $self, @args ) = @_;
  if ( @args and not $args[0] ) {
    $self->_preserve(0);
    $self->_dir->unlink_on_destroy(1);
    return;
  }
  $self->_preserve(1);
  $self->_dir->unlink_on_destroy(0);
  return 1;
}







sub _clean_pkg {
  my ($package) = @_;
  $package =~ s/::/-/gsmx;
  $package =~ s/[^\w-]+/_/gsmx;
  return $package;
}







sub _clean_ver {
  my ($ver) = @_;
  return 'versionundef' if not defined $ver;
  $ver =~ s/[^v\d_.]+/_/gsmx;
  return $ver;
}







sub _build__dir {
  my ($self) = shift;

  my $template = q{perl-};
  $template .= _clean_pkg( $self->package );

  if ( $self->with_version ) {
    $template .= q{-} . _clean_ver( $self->package->VERSION );
  }
  if ( $self->with_timestamp ) {
    $template .= q{-} . time;
  }
  if ( $self->with_pid ) {
    ## no critic ( ProhibitPunctuationVars )
    $template .= q{-} . $$;
  }
  $template .= q{-} . ( 'X' x $self->num_random );

  my $dir = File::Temp->newdir( TEMPLATE => $template, TMPDIR => 1 );
  if ( $self->_preserve ) {
    $dir->unlink_on_destroy(0);
  }
  return $dir;
}









sub dir {
  my ($self) = shift;
  return $self->_dir . q[];
}
















sub run_once_in {
  my ( $self, $options, $code ) = @_;
  $code = $options unless defined $code;
  require File::pushd;
  {
    ## no critic (Variables::ProhibitUnusedVarsStricter)
    my $marker = File::pushd::pushd( $self->dir );
    $code->( $self->dir );
  }

  # Dir POP.
  $self->_clear_dir;
  return $self;
}

no Moo;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

File::Tempdir::ForPackage - Easy temporary directories associated with packages.

=head1 VERSION

version 1.000003

=head1 DESCRIPTION

This is mostly an interface wrapper for C<File::Temp::tempdir>, stealing code from C<File::Tempdir>;

=over 4

=item * I constantly forget how C<File::Tempdir> works

=item * I often want a C<tempdir> with the name of the package working with it enshrined in the path

=item * I constantly forget the magic glue syntax to get a folder inside a system C<tempdir> with a custom prefix and a user defined length of random characters.

=back

And this is designed to solve this simply.

use File::TempDir::ForPackage;

  my $tempdir = File::TempDir::ForPackage->new( package => __PACKAGE__ , use_version => 1 );
  my $dir = $tempdir->dir();

do shit in `$dir`
$dir on Linux will be something like /tmp/perl-Some-Package-maybewith-a-VERSION-AFG14561/
so if it crashes and leaves a C<tempdir> behind, you will know who left that C<tempdir> behind and have a way of cleaning it up.

When C<$tempdir> is destroyed, C<$dir> will be cleaned;

Additionally:

  $dir->run_once_in(sub{
    ...
  });

Is there for people who don't trust scope auto-cleansing and want to know when the directory is reaped.

Additionally, this code can be run in a tight loop creating and destroying lots of similarly named temporary directories without risk of conflict.

  for my $i ( 0 .. 30  ) {
    $dir->run_once_in(sub {
      system 'find $PWD';
    });
  }

This emits something like:

  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-PzH4BD
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-5h8nkG
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-UXKt4S
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-Lqg2aW
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-DkNeq6
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-jRI_zF
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-j0_Gt1
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-iX1ddT
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-ZmvikK
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-QNGOUF
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-6wssvL
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-ZmwZxl
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-wIzRTs
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-xetCym
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-8Y0vyX
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-Zlqt6X
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-U5Z_Sa
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-sKmow1
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-rUND95
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-XjPSGF
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-ec8sZZ
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-_4NBwX
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-xM9i6l
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-p3FhJf
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-Zv0sso
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-rP8cAi
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303496-408662-iade0x
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-fsDDPy
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-FeCcfZ
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-ta5yfg
  /tmp/perl-File-Tempdir-ForPackage-versionundef-1343303497-408662-rdcQhF

Except of course, with a package of your choosing, and possibly that packages version.

=head1 METHODS

=head2 C<preserve>

Toggle the preservation of the C<tempdir> after it goes out of scope or is otherwise indicated for cleaning.

  $instance->preserve(); # tempdir is now preserved after cleanup
  $instance->preserve(0); # tempdir is purged at cleanup
  $instance->preserve(1); # tempdir is preserved after cleanup

Note that in C<run_once_in>, a new C<tempdir> is created and set for this modules consumption for each run of C<run_once_in>,
regardless of this setting. All this setting will do, when set, will prevent each instance being reaped from the file system.

Thus:

  $dir->preserve(1);
  for( 1..10 ){
    $dir->run_once_in(sub{

    });
  }

Will create 10 temporary directories on your file system and not reap them.

=head2 C<dir>

Return a path string to the created temporary directory

  my $path = $instance->dir

=head2 C<run_once_in>

Vivifies a temporary directory for the scope of the passed sub.

  $instance->run_once_in(sub{
    # temporary directory is created before this code runs.
    # Cwd::getcwd is now inside the temporary directory.
  });

  # temporary directory is reset, and possibly reaped.

You can call this method repeatedly, and you'll get a separate temporary directory each time.

=head1 ATTRIBUTES

=head2 C<package>

The package to report as being associated with.
This really can be any string, as its sanitized and then used as a path part.

If not specified, will inspect C<caller>

  my $instance = CLASS->new(
    package => 'Something::Here',
    ...
  );

Note: If you want C<with_version> to work properly, specifying a valid package name will be helpful.

=head2 C<with_version>

Include the version from C<< package->VERSION() >> in the C<tempdir> path.

Defaults to false.

  my $instance = CLASS->new(
    ...
    with_version => 1,
  );

=head2 C<with_timestamp>

Include C<time> in the C<tempdir> path.

Defaults to false.

  my $instance = CLASS->new(
    ...
    with_timestamp => 1,
  );

=head2 C<with_pid>

Include C<$$> in the C<tempdir> path.

Defaults to false.

  my $instance = CLASS->new(
    ...
    with_pid => 1,
  );

=head2 C<num_random>

The number of characters of randomness to include in the C<tempdir> template.

Defaults to 8. Must be no lower than 4.

  my $instance = CLASS->new(
    ...
    num_random => 5,
  );

=head1 PRIVATE ATTRIBUTES

=head2 C<_preserve>

Internal C<boolean> for tracking the _preserve state.

=head2 C<_dir>

Internal C<File::Tempdir> path.

=head1 PRIVATE METHODS

=head2 C<_build__dir>

Builder method for _dir which creates a temporary directory based on the passed parameters.

=head1 PRIVATE FUNCTIONS

=head2 C<_clean_pkg>

Scrape garbage out of the 'package' field for use in file system tokens.

=head2 C<_clean_ver>

Scrape garbage out of versions for use in file system tokens.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
